### PR TITLE
Reduce video reader memory consumption

### PR DESCRIPTION
**Pull request by JanuszL**
- decouples memory occupied by the decoded samples from the prefetching buffer
- moves the decoding of the samples after the batch has been formed in the prefetch method

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It reduces video reader memory consumption by moving the decoding of the samples after the batch has been formed in the prefetch method

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     decouples memory occupied by the decoded samples from the prefetching buffer
     the memory consumption is not proportional to the batch size, not the initial fill for shrading
 - Affected modules and functionalities:
     VideoReader
 - Key points relevant for the review:
     NA
 - Validation and testing:
     current test applies
 - Documentation (including examples):
     NA

**JIRA TASK**: *[NA]*

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

File: dali/operators/reader/video_reader_op.h [hunk 1]
```diff
@@ -86,30 +88,20 @@ class VideoReader : public DataReader<GPUBackend, SequenceWrapper> {
...
-    output.Resize(output_shape);
   }
 
-  void PrepareVideoOutput(TensorList<GPUBackend> &output, DeviceWorkspace &ws) {
-    SetOutputType(output);
-    SetOutputShape(output, ws);
-    output.SetLayout("FHWC");
+ protected:
+  virtual void SetOutputShapeType(TensorList<GPUBackend> &output, DeviceWorkspace &ws) {
+    output.Resize(prefetched_batch_tensors_[curr_batch_consumer_].shape(),
```
Comment (jantonguirao) [inline dali/operators/reader/video_reader_op.h:103]:
```suggestion
    auto &curr_batch = prefetched_batch_tensors_[curr_batch_consumer_];
    output.Resize(curr_batch .shape(), curr_batch.type());
```

File: dali/operators/reader/video_reader_op.h [hunk 2]
```diff
@@ -132,16 +124,11 @@ class VideoReader : public DataReader<GPUBackend, SequenceWrapper> {
...
     TensorList<GPUBackend> &video_output,
-    SequenceWrapper &prefetched_video,
+    TensorList<GPUBackend> &video_batch,
     DeviceWorkspace &ws) {
-    video_output.type().Copy<GPUBackend, GPUBackend>(
-      video_output.raw_mutable_tensor(data_idx),
-      prefetched_video.sequence.raw_data(),
-      prefetched_video.sequence.size(),
-      ws.stream());
+    video_output.Copy(video_batch, ws.stream());
```
Comment (jantonguirao) [inline dali/operators/reader/video_reader_op.h:131]: This indentation (same for args and code) is quite confusing. Can you autoformat?

File: dali/operators/reader/video_reader_op.h [hunk 3]
```diff
@@ -168,18 +155,23 @@ class VideoReader : public DataReader<GPUBackend, SequenceWrapper> {
...
+    auto& curent_batch = prefetched_batch_tensors_[curr_batch_consumer_];
 
-    PrepareVideoOutput(video_output, ws);
+    SetOutputShapeType(video_output, ws);
     PrepareAdditionalOutputs(ws);
 
-    for (int data_idx = 0; data_idx < batch_size_; ++data_idx) {
-      auto& prefetched_video = GetSample(data_idx);
+    ProcessVideo(video_output, curent_batch, ws);
+    video_output.SetLayout("FHWC");
```
Comment (jantonguirao) [inline dali/operators/reader/video_reader_op.h:164]: maybe setting layout should be part of SetOutputShapeType?

File: dali/operators/reader/nvdecoder/sequencewrapper.h [hunk 4]
```diff
@@ -44,11 +44,10 @@ struct SequenceWrapper {
     this->channels = channels;
     this->dtype = dtype;
 
-    TYPE_SWITCH(dtype, type2id, OutputType, SEQUENCEWRAPPER_SUPPORTED_TYPES, (
-        sequence.set_type(TypeInfo::Create<OutputType>());
-      ), DALI_FAIL(make_string("Not supported output type:", dtype));); // NOLINT
+    // TYPE_SWITCH(dtype, type2id, OutputType, SEQUENCEWRAPPER_SUPPORTED_TYPES, (
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/reader/nvdecoder/sequencewrapper.h:47]: Leftover. It no longer have tensor and type of the wrapper is set just above.

File: dali/operators/reader/video_reader_op.h (hunk 2 above)
Comment (JanuszL) [reply dali/operators/reader/video_reader_op.h:131]: Done

File: dali/operators/reader/video_reader_op.h (hunk 1 above)
Comment (JanuszL) [reply dali/operators/reader/video_reader_op.h:103]: Done

File: dali/operators/reader/video_reader_op.h [hunk 5]
```diff
@@ -86,30 +88,20 @@ class VideoReader : public DataReader<GPUBackend, SequenceWrapper> {
...
-    }
-  }
-
-  virtual void SetOutputShape(TensorList<GPUBackend> &output, DeviceWorkspace &ws) {
-    TensorListShape<> output_shape(batch_size_, sequence_dim);
-    for (int data_idx = 0; data_idx < batch_size_; ++data_idx) {
-      output_shape.set_tensor_shape(
-        data_idx, GetSample(data_idx).sequence.shape());
+  inline ~VideoReader() {
+    // wait for all kernels working on the sequences and the underlying memory is kept in
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/reader/video_reader_op.h:92]: Done

File: dali/operators/reader/video_reader_op.cc [hunk 6]
```diff
@@ -23,6 +23,29 @@
...
+  // resize the current batch
+  std::vector<TensorShape<>> tmp_shapes;
+  auto ref_type = curr_batch[0]->dtype;
+  for (auto &sample : curr_batch) {
+    DALI_ENFORCE(ref_type == sample->dtype,
+                  "The tensors in the input batch do not all have the same type");
+    tmp_shapes.push_back(sample->shape());
+  }
+
+  curr_tensor_list.Resize(TensorListShape<>(tmp_shapes), TypeTable::GetTypeInfo(ref_type));
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/reader/video_reader_op.cc]: Done

File: dali/operators/reader/video_reader_op.cc [hunk 7]
```diff
@@ -23,6 +23,29 @@
...
+  auto &curr_batch = prefetched_batch_queue_[curr_batch_producer_];
+  auto &curr_tensor_list = prefetched_batch_tensors_[curr_batch_producer_];
+
+  // resize the current batch
+  std::vector<TensorShape<>> tmp_shapes;
+  auto ref_type = curr_batch[0]->dtype;
+  for (auto &sample : curr_batch) {
+    DALI_ENFORCE(ref_type == sample->dtype,
+                  "The tensors in the input batch do not all have the same type");
+    tmp_shapes.push_back(sample->shape());
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/reader/video_reader_op.cc]: Done

File: dali/operators/reader/video_reader_op.cc [hunk 8]
```diff
@@ -23,6 +23,29 @@
...
+  DataReader<GPUBackend, SequenceWrapper>::Prefetch();
+  auto &curr_batch = prefetched_batch_queue_[curr_batch_producer_];
+  auto &curr_tensor_list = prefetched_batch_tensors_[curr_batch_producer_];
+
+  // resize the current batch
+  std::vector<TensorShape<>> tmp_shapes;
+  auto ref_type = curr_batch[0]->dtype;
+  for (auto &sample : curr_batch) {
+    DALI_ENFORCE(ref_type == sample->dtype,
+                  "The tensors in the input batch do not all have the same type");
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/reader/video_reader_op.cc]: Done

File: dali/operators/reader/video_reader_op.cc [hunk 9]
```diff
@@ -23,6 +23,29 @@
 
 namespace dali {
 
+void VideoReader::Prefetch() {
+  DataReader<GPUBackend, SequenceWrapper>::Prefetch();
+  auto &curr_batch = prefetched_batch_queue_[curr_batch_producer_];
+  auto &curr_tensor_list = prefetched_batch_tensors_[curr_batch_producer_];
+
+  // resize the current batch
+  std::vector<TensorShape<>> tmp_shapes;
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/reader/video_reader_op.cc]: Done

File: dali/operators/reader/video_reader_op.h (hunk 3 above)
Comment (JanuszL) [reply dali/operators/reader/video_reader_op.h]: Done

File: dali/operators/reader/video_reader_op.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/reader/video_reader_op.cc]: I tried but for some reason it doesn't work. Changing back to `std::vector<TensorShape<>>` fixes crash.

**Comment by JanuszL** [comment]
!build

File: dali/operators/reader/nvdecoder/sequencedesc.h [hunk 10]
```diff
@@ -92,22 +87,17 @@ struct SequenceWrapper {
...
   TensorShape<3> frame_shape() const {
     return TensorShape<3>{height, width, channels};
   }
 
-  Tensor<GPUBackend> sequence;
+  TensorShape<4> shape() const {
+    return TensorShape<4>{count, height, width, channels};
+  }
+
+  // non-owning tensor that should only share data allocated somwhere else
```
Comment (mzient) [inline dali/operators/reader/nvdecoder/sequencedesc.h:98]: Is it really true from this sturcture's perspective? It's used like this, but would anything break if it was used differently?

File: dali/operators/reader/nvdecoder/sequencedesc.h [hunk 11]
```diff
@@ -29,12 +29,12 @@
 
 namespace dali {
 
-#define SEQUENCEWRAPPER_SUPPORTED_TYPES (float, uint8_t)
+#define SequenceDesc_SUPPORTED_TYPES (float, uint8_t)
 
 // Struct that Loader::ReadOne will read
-struct SequenceWrapper {
+struct SequenceDesc {
```
Comment (mzient) [inline dali/operators/reader/nvdecoder/sequencedesc.h:35]: If the data buffer _is_ still here, then I don't think the renaming is warranted.

File: dali/operators/reader/video_reader_op.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (mzient) [reply dali/operators/reader/video_reader_op.cc:38]: How can this happen? Is that really a matter for DALI_ENFORCE or assert  / logic error?

File: dali/operators/reader/video_reader_op.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/reader/video_reader_op.cc:32]: tensor list shape doesn't work for me for some reason.

File: dali/operators/reader/video_reader_op.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/reader/video_reader_op.cc:38]: It should be assert.

File: dali/operators/reader/nvdecoder/sequencedesc.h (hunk 10 above)
Comment (JanuszL) [reply dali/operators/reader/nvdecoder/sequencedesc.h]: It should still work, from this class perspective.

File: dali/operators/reader/nvdecoder/sequencedesc.h (hunk 11 above)
Comment (JanuszL) [reply dali/operators/reader/nvdecoder/sequencedesc.h]: Reverted.

**Review by mzient — APPROVED** [review]

File: dali/operators/reader/nvdecoder/sequencewrapper.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/operators/reader/nvdecoder/sequencewrapper.h]: Now it has tensor back, but the type is set when it shares the memory with other Tensor.

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Review by jantonguirao — APPROVED** [review]

**Comment by Timen** [comment]
@JanuszL I can confirm this PR greatly reduced memory consumption, I'm now able to train with large batch size and sequence length when using the VideoReaderResize 👍

**Comment by Timen** [comment]
@JanuszL I build this locally to test, but I just wanted to point out that there hasn't been a nightly build since October the 6th as can be seen here: https://developer.download.nvidia.com/compute/redist/nightly/nvidia-dali-nightly-cuda100/ Did the nightly build not get triggered?

**Comment by JanuszL** [comment]
> @JanuszL I build this locally to test, but I just wanted to point out that there hasn't been a nightly build since October the 6th as can be seen here: https://developer.download.nvidia.com/compute/redist/nightly/nvidia-dali-nightly-cuda100/ Did the nightly build not get triggered?

Nightly builds are published only when they successfully pass the test. If there is a failure it is not published.

**Comment by Timen** [comment]
Is there way for outside contributors/users to see the output of the nightly builds? If not the complete log than maybe a status indicator on the main page reporting build pass/fail could be helpful.

**Comment by JanuszL** [comment]
> @JanuszL I build this locally to test, but I just wanted to point out that there hasn't been a nightly build since October the 6th as can be seen here: https://developer.download.nvidia.com/compute/redist/nightly/nvidia-dali-nightly-cuda100/ Did the nightly build not get triggered?

Nightly builds are published only when they successfully pass the test. If there is a failure it is not published.

**Comment by JanuszL** [comment]
Currently it is not possible as our CI is fully internal.

**Pull request merged**